hashicorp/consul-k8s
# cli: support install command

(This description is lightly edited from the original PR Saad created: https://github.com/hashicorp/consul-k8s-cli/pull/2 We migrated the PR here and migrated some of the review comments as well)

Changes proposed in this PR:
 
* `consul-k8s install` installs the latest Helm chart onto your Kubernetes cluster after performing a few verification steps. Note that once the CLI is moved into the monorepo, the version of the Helm chart deployed will match the version in the repository.
* For details about flags and their intended behavior visit the RFC, linked [here](https://docs.google.com/document/d/13vBeED9OkKUO3isZWIxTRvbH-sxEzvYH9rdJeb9my48/edit#).
* `common/flag` is copied from Waypoint’s internal/pkg subdirectory [here](https://github.com/hashicorp/waypoint/tree/main/internal/pkg/flag). It is a thin layer over the standard Go flag package that provides features such as aliasing and global flags. It was created specifically for mitchellh/cli. See also [this](https://hashicorp.slack.com/archives/C0E7S0EU8/p1628019710015500) Slack thread for the decision to use it.
* `common/terminal` package is adapted from [the Waypoint Plugin SDK](https://github.com/hashicorp/waypoint-plugin-sdk/tree/74d9328929293551499078da388b8d057f3b2341/terminal) and excludes the Status interface and Step interface as described in [doc.go](https://github.com/hashicorp/consul-k8s-cli/blob/install/cmd/common/terminal/doc.go). It allows us to have a consistent UI output with style.
* `common/base.go` describes a base command that can be embedded into all commands, and pass in common functionality like UI and ctx. This pattern was adapted from Waypoint.
* ~`common/utils.go` contains a function `WithInterrupt()` that we take from [signalcontext.go](https://github.com/hashicorp/waypoint/blob/348d2c77fce199952618ccef6433c8844b22583b/internal/pkg/signalcontext/signalcontext.go) in Waypoint's internal/pkg.~
 
How I've tested this PR:
 
* Created rather simple unit tests that validate two different helper functions inside of install.go. The first helper function we test is `validateFlags()` which performs a lot of the basic sanity checks before starting the install. The second helper is `validLabel()` which makes sure that the namespace the user passes is legal, in that if follows [RFC 1123 label convention](https://kubernetes.io/docs/concepts/overview/working-with-objects/names/).
* In addition to the actual unit tests, a lot of manual ad-hoc testing took place. Follow the instructions below to build the CLI and test the various flags, especially the -set ones, and see how they interact with one another.
* Note that more rigorous acceptance tests are soon to come after monorepo is finished. These tests will actually run the installation through and validate it worked properly.
 
How I expect reviewers to test this PR:
 
* Try it out :
       `git clone https://github.com/hashicorp/consul-k8s-cli.git`
       `cd consul-k8s-cli`
       `go build -o consul-k8s`
       `consul-k8s install`
 
* Try this with different flags and various configurations. Note that if you are running on minikube, it is recommended to install with `-preset=demo`. That way server.replicas is set to 1. I also found it helpful to specify the namespace when manually testing via `-namespace=not-peppertrout`.
 
* Code review :
       Pay careful attention to install.go, specifically the pre-install checks. If you see anything you believe should be pulled out into an individual function and/or tested more thoroughly please comment!

Checklist:
  > HashiCorp engineers only, community PRs should not add a changelog entry.
  > Entries should use present tense (e.g. Add support for...)
  
Co-authored-by: Saad Jamal [saad@hashicorp.com](mailto:saad@hashicorp.com)
Co-authored-by: Nitya Dhanushkodi [nitya@hashicorp.com](mailto:nitya@hashicorp.com)
